pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/intervalST/IntervalST.java

```diff
@@ -0,0 +1,197 @@
+      }
+    }
+
+    _maxs = new ArrayList<>(_intervals.size());
+    buildAuxiliaryInfo();
+  }
+
+  private void buildIntervalTree(List<Interval> sortedIntervals, List<Interval> res) {
+    // Build interval bst by bfs, the root for each subtree will be the median interval.
+    // A typical balanced tree
+    //                              [10, 20]
+    //                              /       \
+    //                       [8, 15]        [12, 20]
+    //                          /            /
+    //                   [5, 10]       [10, 30]
+    //
+    // will be represented as { [10, 20], [8, 15], [12, 20], [5, 10], null, [10, 30] }
+    if (sortedIntervals.size() == 0) {
+      return;
+    }
+
+    LinkedList<Integer> startIndexQueue = new LinkedList<>();
+    LinkedList<Integer> endIndexQueue = new LinkedList<>();
+    startIndexQueue.addLast(0);
+    endIndexQueue.addLast(sortedIntervals.size());
+    int count = 0;
+    while (!startIndexQueue.isEmpty() && count < sortedIntervals.size()) {
+      int start = startIndexQueue.pollFirst();
+      int end = endIndexQueue.pollFirst();
+      if (start < end) {
+        int mid = start + (end - start) / 2;
+        res.add(sortedIntervals.get(mid));
+        count++;
+        startIndexQueue.addLast(start);
+        endIndexQueue.addLast(mid);
+        startIndexQueue.addLast(mid + 1);
+        endIndexQueue.addLast(end);
+      } else {
+        res.add(null);
+      }
+    }
+  }
+
+  private void buildAuxiliaryInfo() {
+    // Initialize and build max info for the interval tree by dfs
+    for (int i = 0; i < _intervals.size(); i++) {
+      _maxs.add(Long.MIN_VALUE);
+    }
+    buildAuxiliaryInfo(0);
+  }
+
+  private void buildAuxiliaryInfo(int nodeIndex) {
+    if (!hasNode(nodeIndex)) {
+      return;
+    }
+
+    int leftChildIndex = getLeftChildIndex(nodeIndex);
+    int rightChildIndex = getRightChildIndex(nodeIndex);
+
+    buildAuxiliaryInfo(leftChildIndex);
+    buildAuxiliaryInfo(rightChildIndex);
+
+    long max = _intervals.get(nodeIndex)._max;
+    max = Math.max(getMax(rightChildIndex), Math.max(max, getMax(leftChildIndex)));
+    _maxs.set(nodeIndex, max);
+  }
+
+  private int getLeftChildIndex(int nodeIndex) {
+    return nodeIndex * 2 + 1;
+  }
+
+  private int getRightChildIndex(int nodeIndex) {
+    return  nodeIndex * 2 + 2;
+  }
+
+  private long getMax(int index) {
+    if (!hasNode(index)) {
+      return Long.MIN_VALUE;
+    }
+    return _maxs.get(index);
+  }
+
+  /**
+   * Find all values whose intervals intersect with the input interval.
+   *
+   * @param searchInterval search interval
+   * @return list of all qualified values.
+   */
+  public List<Value> searchAll(Interval searchInterval) {
+    List<Value> list = new ArrayList<>();
+    if (searchInterval == null) {
+      return list;
+    }
+    searchAll(0, searchInterval, list);
+    return list;
+  }
+
+  private boolean searchAll(int nodeIndex, Interval searchInterval, List<Value> list) {
```

Comment:
I think we can simplify the algorithm as followings:
```
- Check root
- If (left.max >= interval.min) search left
- If (root.min <= interval.max) search right
```
Then we don't need to maintain these extra flags, and it is actually more efficient because if `left.max >= interval.min` and you cannot find intersections on left subtree, then `root.min` is always larger than `interval.max`

Reply:
Yes, I agree.